### ISSUE TITLE
Add native yescrypt support

### ISSUE THREAD

**Issue by solardiz** [openwall/john#4621]
Although we have and use yescrypt in the tree, we currently only use it to provide scrypt (for several formats), but not native yescrypt. For Linux distros that have started to use native yescrypt hashes via libxcrypt (ALT Linux, Kali Linux) and for those that support them (Fedora 29+), these hashes can currently be cracked by using `--format=crypt`. However, this is suboptimal (memory is getting (de)allocated all the time, which slows things down) and non-portable (the hashes cannot be cracked on a system that doesn't have yescrypt support in its libcrypt). We should add the support into JtR itself, possibly in the same format with scrypt. (A next task then would be also supporting yescrypt ROM.)

**Comment by roycewilliams** [comment]
Distro update: Debian 11 (bullseye) now defaults to yescrypt.

https://www.debian.org/releases/bullseye/amd64/release-notes/ch-information.en.html#pam-default-password

**Comment by solardiz** [comment]
> Distro update: Debian 11 (bullseye) now defaults to yescrypt.

Right, and Fedora 35 likely will, too (now in beta).

**Comment by magnumripper** [comment]
Not having proper support for `$y$` in JtR is embarrassing, considering who's the author of both.

I wouldn't know where to start. Would it work on GPU at all?

**Comment by solardiz** [comment]
Yes, embarrassing it is. I even thought about allocating some of my time to it this year, but ended up taking care of other priorities. The CPU format should be fairly easy, it's just that I wanted to implement a universal parameters decode in upstream yescrypt and use that. But we can also implement a hack first.

Yes, we also need it in OpenCL, and yes it will work on GPU - but slowly. We primarily need it to have proper comparisons vs. CPU and vs. Argon2, and to inform future decisions on parameters to use defensively.

**Comment by solardiz** [comment]
Oh, as to where to start with OpenCL code - it's the yescrypt 0.5'ish PoW miners. yescrypt 1.0 has changed (and should be more GPU-unfriendly), but the overall approach is the same. See the last couple of messages I posted to https://lists.openwall.net/phc-discussions/ in 2018.
